### PR TITLE
Add error handling for delayed responses

### PR DESCRIPTION
**Pull request by xsawyerx**
GH #941 mentions callbacks for delayed responses. Currently the
biggest one is error handling. This is what this commit adds.

It does this by introducing another argument to the "delayed"
keyword, "on_error".

    get '/' => sub {
        delayed {

        } on_error => sub {

        };
    };

The "on_error" option receives a sub. Since the "delayed" request
is already done within the context of a delayed response, there's
no need to use "delayed" again in the sub, just providing a regular
callback instead.

The tests don't fully work since the development server, for some
still unbeknownst reason, doesn't actually crash after we close
the socket and write again. In Twiggy it does. You can use this
test script which uses Twiggy:

    use strict;
    use warnings;
    my $caught_error;

    {
        package App::ErrorHandler; ## no critic
        use Dancer2;
        use AnyEvent;
        get '/log' => sub {
            delayed {
                flush;
                content "ping\n";
                done;
                content "failure\n";
            };
        };

        get '/cb' => sub {
            delayed {
                flush;
                content "ping\n";
                done;
                content "failure\n";
            } on_error => sub {
                my $caught_error = shift;
                core "Failed to stream to user: " . request->remote_address;
                1;
            };
        };
    }

    App::ErrorHandler->to_app;

Original tests marked as TODO until they work.

### CODE REVIEW

File: lib/Dancer2/Core/App.pm [hunk 1]
```diff
@@ -889,7 +889,11 @@ sub send_file {
                 [ $res->status, $res->headers_to_array, $fh ]
             );
         };
+
+        Scalar::Util::weaken( my $weak_self = $self );
+
         $response = Dancer2::Core::Response::Delayed->new(
+            error_cb => sub { $weak_self->logger_engine->log(@_) },
```
Comment (veryrusty) [inline lib/Dancer2/Core/App.pm]: The call to `logger_engine->log(@_)` is missing an argument. It needs both a *log level* and the log message.

**Comment by veryrusty** [comment]
This looks good. @xsawyerx++
I'll take a final pass over it tomorrow. :)

File: lib/Dancer2/Core/App.pm (hunk 1 above)
Comment (xsawyerx) [inline lib/Dancer2/Core/App.pm]: Can you add a test plus a fix? Pretty please? :)

**Comment by xsawyerx** [comment]
@veryrusty Thank you. :)

File: lib/Dancer2/Core/App.pm (hunk 1 above)
Comment (xsawyerx) [inline lib/Dancer2/Core/App.pm]: Pinging again on this, @veryrusty, I think I tested it when committing it and it worked. Basically when you call log, it just passes it on to the logger. Are you sure this is a mistake?

Comment (veryrusty) [inline lib/Dancer2/Core/App.pm]: Sorry to go MIA on this. I'm sure its missing an arg.  
  * `error_cb` takes the error message as its only argument. 
  * The `log` method on all the logging engines take two args: "msg_level", and  "msg_content".

The error callback added to the `content` keyword in this PR  [two args for its logging call](https://github.com/PerlDancer/Dancer2/blob/feature/delayed-on-error/lib/Dancer2/Core/DSL.pm#L301-L303)

This is inside of `send_file`, which uses delayed responses to send a filehandle back to the PSGI engine. So `$Dancer2::Core::Route::WRITER->write` shouldn't be called in this use case; I'm not certain its possible to write a test case for it (at least I can't think of how at the moment).
 [ The fix would be: `error_cb => sub { $weak_self->logger_engine->log(warning => @_) }` ]

Comment (xsawyerx) [inline lib/Dancer2/Core/App.pm]: I'll amend it to contain the fix. Thanks! Unfortunately the development server's failure is kind of weird. I couldn't write a proper test for it.

**Comment by xsawyerx** [comment]
@veryrusty Added the fix! Thanks!

I'll be merging this soon.

**Pull request closed without merging**